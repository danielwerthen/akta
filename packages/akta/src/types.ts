import { Observable } from 'rxjs';

export type AktaElement<
  P extends { [key: string]: unknown } = { [key: string]: unknown }
> = {
  $$type: Symbol;
  type: string | AktaComponent<P>;
  props: P;
  key: string | undefined;
};

export type ObservableElements =
  | Observable<AktaAllElements>
  | Generator<AktaAllElements, AktaAllElements>
  | AsyncGenerator<AktaAllElements, AktaAllElements>;

export type AktaAllElements = AktaElement | ObservableElements;

export type AktaNode = AktaAllElements | string | null | AktaNode[];

export const AktaElementType = Symbol('Akta element');

export function isAktaElement(node: unknown): node is AktaElement {
  if (typeof node === 'object') {
    return (node as AktaElement)?.$$type === AktaElementType;
  }
  return false;
}

export interface ElementProperties {}

export type AktaComponent<PROPS extends {}> = (
  props: PROPS
) => AktaAllElements | string | null;

export function AktaPreparedComponent(_props: {
  children: Observable<HTMLElement>;
}) {
  return null;
}
