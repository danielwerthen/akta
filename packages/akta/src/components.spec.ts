import { mount, ErrorBoundary, jsx } from './index';
import { Subject } from 'rxjs';
import { AktaNode } from './types';

describe('ErrorBoundary', () => {
  it('should handle component exceptions gracefully', async () => {
    const root = document.createElement('div');
    function ErrorImmediately() {
      throw new Error('Should be caught');
      return 'Never';
    }
    mount(
      jsx(ErrorBoundary, {
        fallback: 'fallback',
        children: jsx('div', {
          children: [jsx(ErrorImmediately, {})],
        }),
      }),
      root
    );
    expect(root).toMatchSnapshot();
  });
  it('should handle eventual component exceptions gracefully', async () => {
    const root = document.createElement('div');
    const subject = new Subject<AktaNode>();
    function ErrorLater() {
      return subject;
    }
    const propSubject = new Subject<AktaNode>();
    function ErrorProp() {
      return jsx('div', { className: propSubject });
    }
    mount(
      [
        jsx(ErrorBoundary, {
          fallback: 'fallback',
          children: jsx('div', {
            children: [jsx(ErrorLater, {})],
          }),
        }),
        jsx(ErrorBoundary, {
          fallback: 'fallback2',
          children: jsx('div', {
            children: [jsx(ErrorProp, {})],
          }),
        }),
      ],
      root
    );
    expect(root).toMatchSnapshot();
    subject.next('It works');
    propSubject.next('className');
    expect(root).toMatchSnapshot();
    subject.error(new Error('Foobar'));
    propSubject.next('className2');
    expect(root).toMatchSnapshot();
    propSubject.error(new Error('Foobar'));
    expect(root).toMatchSnapshot();
  });
});
