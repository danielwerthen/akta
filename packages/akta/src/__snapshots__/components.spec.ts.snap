// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary should handle component exceptions gracefully 1`] = `
<div>
  fallback
</div>
`;

exports[`ErrorBoundary should handle eventual component exceptions gracefully 1`] = `<div />`;

exports[`ErrorBoundary should handle eventual component exceptions gracefully 2`] = `
<div>
  <div>
    It works
  </div>
  <div>
    <div
      class="className"
    />
  </div>
</div>
`;

exports[`ErrorBoundary should handle eventual component exceptions gracefully 3`] = `
<div>
  fallback
  <div>
    <div
      class="className2"
    />
  </div>
</div>
`;

exports[`ErrorBoundary should handle eventual component exceptions gracefully 4`] = `
<div>
  fallback
  fallback2
</div>
`;
